### ISSUE TITLE
Metric counts are based on arbitrary, discrete limits

### ISSUE THREAD

**Issue by EdvardM** [bbatsov/rubocop#1753]
I apologize if this is kinda out of scope -issue. I know definitely Rubocop itself is not to blame. But author being a smart guy I know there's a way to circumvent the madness described below.

If anyone has listened to Sandi Metz's speak about programming, she must have heard of complexity metrics. While useful, they often suffer from this -- how should I put it politely -- arbitrary, binary rule; a given piece code is good fi it contains less than k lines, and bad afterwards. Be it a method, class of file. 

We all know such metrics are arbitrary. Even if you find a nice ballpark figure that is suitable for your project.. well, seriously: say you have defined 100 lines to be maximum for a class. That is, code with 100 lines is good, and suddenly at 101 it gets so bad you are notified of the fact. Adjusting the limit doesn't help, obviously. Make the number large enough, and you no longer have any use for that metric. 

My suggestion: instead of binary good/bad score, have a metric that goes exponentially worse with number of lines. That way you are never 'green', but you'd have a useful figure. A simple metric could raise 1.05 to the number of lines, and something similar for class/file. That way you'd keep continuously minimizing the number for top n offenders without arbitrary limits. How would that sound?

**Comment by mattjmcnaughton** [comment]
I like how this idea would allow you to keep on improving your code even if you technically pass all of the rubocop checks. So if you wanted to continue cleaning up and refactoring, you would know some good places to start. Could be a command line option? But also broadens the scope of rubocop, so that's an potential concern.

**Comment by EdvardM** [comment]
Thanks, Matt. Yup, it would broaken scope of Rubocop :( and I dislike incoherence in code. But current metrics are next to useless, to put it bluntly. In an ideal situation I'd have that metric as with the rest, but report is as extra info (ie. not en error). Maybe even have the exponent/base configurable. I guess I need to take a look at the code.

**Comment by jonas054** [comment]
Here's how I use the current solution to work with refactoring the RuboCop source code.

I start by taking a look at [.rubocop_todo.yml](https://github.com/bbatsov/rubocop/blob/master/.rubocop_todo.yml). The `Max` numbers and the comments there help me select a suitable place to tackle. For example, right now I can see that the offense count for `Metrics/ClassLength` is low, so trying to lower that `Max` number seems like a good idea. The goal of removing that cop from the todo list might be attainable.

I then lower `Metrics/ClassLength: Max` by hand, perhaps only by `1`, and run `rubocop` to see which classes are now above the threshold. If I can find a refactoring to make those classes smaller, I do the implementation, run `rubocop --auto-gen-config` and commit all the changes, including the generated ones in `.rubocop_todo.yml`.

So you see I can
* work with continuously lowering the different metrics for the top offenders
* lower them *under* the default values if I want to

It follows that I need to know in more detail (let's say from the point of view of working with refactorings within the RuboCop source code)
* What's wrong with the current way of working?
* What's better with @EdvardM's suggestion? What does the exponential score bring to the table?

**Comment by EdvardM** [comment]
Jonas -- my main point is not that much of exponentiation, but analogue nature of code quality compared to discrete (binary actually) rating. To be fair, I like your solution. It's an excellent solution to my problem in a sense that it requires no changes to code. I just find it awkward to need to modify a file repeatedly. We are dealing with computers, man :) That's just awkward (but like I said, good enough that I might end up using just that). 

My main gripe is still with the actual metric. A code that is without fault at n lines, n being sufficiently greater than one, cannot become bad at n+1. And that's exactly what those metrics imply.

**Comment by jonas054** [comment]
It's the same in traffic. Drive 90 km/h and everything is fine. 91 and you're breaking the law. I don't have a problem with that -- in traffic or in RuboCop -- but just in case you want to explore the possibility of listing metrics offenses in numeric order, a (custom) formatter is the way to go.

[Here](https://gist.github.com/jonas054/3a69860e7e9fcd37aa0f) is a pretty crude example of how that might look.

Put it in your home directory for example, and run with

```
rubocop --require ~/top_offenders_formatter.rb --format TopOffendersFormatter
```

**Comment by EdvardM** [comment]
Good analogue, and I do find that a problem :) Even some police I've talked with acknowledge that there are many other things that would better indicate dangerous driving, but speed limits are what they are because they are easy to monitor. But now that you mentioned it, even with speed limits you get different "score" (say, sanctions) depending on by how much you exceed them. If you go 90mph in 60mph area, you are fined more than when you go 65mph. But I digress.

And thanks for the tip! I'll look at custom formatter. It does seem a more appropriate solution to my itch than the hack I rustled up.
